src/main/java/org/elasticsearch/indices/IndicesService.java

```diff
@@ -419,4 +418,128 @@ public synchronized void beforeIndexShardClosed(ShardId shardId, @Nullable Index
+        } catch (IOException e) {
+            logger.warn("[{}] failed to delete closed index", e, metaData.index());
+        }
+    }
+
+    /**
+     * Deletes the index store trying to acquire all shards locks for this index.
+     * This method will delete the metadata for the index even if the actual shards can't be locked.
+     */
+    public void deleteIndexStore(String reason, ClusterState clusterState, IndexMetaData metaData) throws IOException {
+        if (nodeEnv.hasNodeFile()) {
+            String indexName = metaData.getIndex();
+            if (clusterState.metaData().hasIndex(indexName)) {
+                final IndexMetaData index = clusterState.metaData().index(indexName);
+                throw new ElasticsearchIllegalStateException("Can't delete index store for [" + indexName + "] - it's still part of the cluster state [" + index.getUUID() + "] [" + metaData.getUUID() + "]");
+            }
+            Index index = new Index(indexName);
+            final Settings indexSettings = buildIndexSettings(metaData);
+            deleteIndexStore(reason, index, indexSettings);
+        }
+    }
+
+    private void deleteIndexStore(String reason, Index index, Settings indexSettings) throws IOException {
+        try {
+            // we are trying to delete the index store here - not a big deal if the lock can't be obtained
+            // the store metadata gets wiped anyway even without the lock this is just best effort since
+            // every shards deletes its content under the shard lock it owns.
+            logger.info("deleting index store reason [{}]", reason);
+            nodeEnv.deleteIndexDirectorySafe(index, 0, indexSettings);
+        } catch (LockObtainFailedException ex) {
+            logger.debug("[{}] failed to delete index store - at least one shards is still locked", ex, index);
+        } catch (Exception ex) {
+            logger.warn("[{}] failed to delete index", ex, index);
+        } finally {
+            // this is a pure protection to make sure this index doesn't get re-imported as a dangeling index.
+            // we should in the future rather write a tombstone rather than wiping the metadata.
+            MetaDataStateFormat.deleteMetaState(nodeEnv.indexPaths(index));
+        }
+    }
+
+    /**
+     * Deletes the shard with an already acquired shard lock. Note
+     * @param reason the reason for the shard deletion
+     * @param lock the lock of the shard to delete
+     * @param indexSettings the shards index settings.
+     * @throws IOException if an IOException occurs
+     */
+    public void deleteShardStore(String reason, ShardLock lock, Settings indexSettings) throws IOException {
+        ShardId shardId = lock.getShardId();
+        if (canDeleteShardContent(shardId, indexSettings) == false) {
+            throw new ElasticsearchIllegalStateException("Can't delete shard " + shardId);
+        }
+        logger.info("{} deleting shard reason [{}]", shardId, reason);
```

Comment:
same here - I think it's better to log the info message if the deletion was successful.